Add slider ticks.

Damnae opened the pull request.

Damnae committed:
Add slider ticks.

Damnae committed:
Cache drawable ticks.

Damnae committed:
Fix slider ticks expiring before scoring happens.

Tom94 commented on osu.Game.Modes.Osu/Objects/Drawables/SliderTicksRenderer.cs:
```diff
@@ -0,0 +1,96 @@
+
+namespace osu.Game.Modes.Osu.Objects.Drawables.Pieces
+{
+    public class SliderTicksRenderer : Container<DrawableSliderTick>
```
I wouldn't call this "Renderer" since rendering is usually the act of drawing itself. `SliderTicksContainer` sounds better to me.

Tom94 commented on osu.Game.Modes.Osu/Objects/Slider.cs:
```diff
@@ -25,17 +28,66 @@ public override int StackHeight
+            var multipliedStartBeatLength = beatmap.BeatLengthAt(StartTime, true);
+
+            Velocity = 100 / multipliedStartBeatLength * baseDifficulty.SliderMultiplier;
+            TickDistance = (100 * baseDifficulty.SliderMultiplier) / baseDifficulty.SliderTickRate / (multipliedStartBeatLength / startBeatLength);
```
Can we please not use / without parens? Also, this whole expression can easily be refactored to be less confusing. For instance by simply multiplying with `startBeatLength`. Lastly I recommend putting `/` to be the _last_ operator in the expression; otherwise someone who does not know operator ordering by heart will read something ambiguous.

Example:

2 / 2 * 4 = ?
(2 / 2) * 4 = 4
2 / (2 * 4) = 1/4

Tom94 replied:
Small note: if you are doing double divisions (like here with `startBeatLength`) you can actually also introduce bugs unrelated to readability. Suppose `startBeatLength == 0`.

`a / (1 / startBeatLength) == NaN` whereas `a * startBeatLength = 0`

Damnae committed:
Merge branch 'master' into slider_ticks

Conflicts:
	osu.Game.Modes.Osu/Objects/Drawables/DrawableSlider.cs
	osu.Game.Modes.Osu/Objects/Slider.cs

Damnae committed:
Clearer slider Velocity and TickDistance calculations.

Damnae committed:
Rename SliderTicksRenderer to SliderTicksLayer.

peppy committed:
Merge branch 'master' into slider_ticks

peppy committed:
Handle case where TickDistance is zero (don't draw ticks).

peppy committed:
Add a TickDistance to TestCaseHitObjects.

peppy committed:
Adjust tick appear/disappear animations.

peppy committed:
Fix namespace.

peppy committed:
Add tick sample.

Tom94 commented on osu.Game.Modes.Osu/Objects/Slider.cs:
```diff
@@ -53,11 +54,13 @@ public override void SetDefaultsFromBeatmap(Beatmap beatmap)
 
-            Velocity = 100 / multipliedStartBeatLength * baseDifficulty.SliderMultiplier;
-            TickDistance = (100 * baseDifficulty.SliderMultiplier) / baseDifficulty.SliderTickRate / (multipliedStartBeatLength / startBeatLength);
+            Velocity = baseVelocity / (timingPoint.BeatLength * velocityAdjustment);
```
Can divide `baseVelocity` by `velocityAdjustment`. Then you don't need to divide by it twice when setting `Velocity` and `TickDistance`.

peppy committed:
Add the concept of nested DrawableHitObjects.

- Applies to Slider Ticks and start circle. repeat/endpoints still need addressing.
- Removed SliderTicksLayer abstraction for now.

Damnae commented on osu.Game.Modes.Osu/Objects/Drawables/DrawableSliderTick.cs:
```diff
@@ -70,8 +69,13 @@ protected override void PlaySample()
-                Judgement.Result = ShouldHit ? HitResult.Hit : HitResult.Miss;
+            {
+                j.Result = Tracking ? HitResult.Hit : HitResult.Miss;
+                j.Score = Tracking ? OsuScoreResult.SliderTick : OsuScoreResult.Miss;
```
using OsuScoreResult.Miss here will count for a 300 point miss instead of 10

peppy replied:
nice catch.

peppy committed:
Add MaxScore to correctly account for the maximum attainable score of DrawableHitObjects.

Tom94 commented on osu.Game/Modes/Objects/Drawables/DrawableHitObject.cs:
```diff
@@ -85,6 +84,19 @@ protected override void LoadComplete()
             Expire(true);
         }
 
+        private List<DrawableHitObject> nestedHitObjects;
```
This seems redundant. Why not make it `Container<DrawableHitObject>` and let it be in the draw hierarchy? Then you don't need to add slider ticks to 2 lists.

peppy replied:
Because this isn't just for sliders. For the time being as sliders have the custom positioning logic I would rather not implement this here just yet.

Also gotta consider potential draw depth issues.

Tom94 replied:
Fair enough, let's discuss this more in next pair programming session
